Sources/NIOHTTP1/HTTPDecoder.swift

```diff
@@ -437,6 +464,50 @@ public class HTTPDecoder<HTTPMessageT>: ByteToMessageDecoder, AnyHTTPDecoder {
             ctx.close(promise: nil)
         }
     }
+
+    private func readEOF(ctx: ChannelHandlerContext) {
+        guard self.state.currentError == nil else {
+            // We're in readEOF because we hit an error and closed the connection.
+            // No need to do this dance again, just return.
+            return
+        }
+
+        guard !self.state.seenEOF else {
+            // We're in readEOF but we have already processed it once for this connection.
+            // Probably this is a channelInactive after half closure, but either way we
+            // shouldn't tell http_parser about this again or it'll send us on_message_complete
+            // again.
+            return
+        }
+
+        // EOF is semantic in HTTP, so we need to tell the parser that we saw it.
+        // We need to be a bit careful: the parser can call out to us, but should only
+        // ever call on_message_complete, which won't do a read. As a result, it *should*
+        // be totally safe to call this with a null data pointer. Just to make sure, though,
+        // let's store nil in the base address. Once we've called this, we've seenEOF: don't
+        // let us enter this function again.
+        self.state.baseAddress = nil
+        self.state.seenEOF = true
+        _ = c_nio_http_parser_execute(&self.parser, &self.settings, nil, 0)
+
+        // We don't need the cumulation buffer, if we're holding it.
+        self.cumulationBuffer = nil
+
+        // No check to state.currentError becuase, if we hit it before, we already threw that
+        // error. This never calls any of the callbacks that set that field anyway. Instead we
+        // just check if the errno is set and throw.
+        let errno = parser.http_errno
```

Comment:
I'm not a huge fan of naming variables `errno` because if you rename them and forget the usage sites, it'll still work. Mind making it `httpError` or `httpErrno` or something?